File: 1013-fibonacci-number/fibonacci-number.cpp
class Solution {
public:
    int fib(int n) {
        // recursion.
        /* if (n == 0)
            return 0;
        if (n == 1)
            return 1;
        return fib(n - 1) + fib(n - 2); */
        int f0 = 0;
        int f1 = 1;
        int f = 0;
        for (int i = 1; i <= n; i++) {
            f0 = f1;
            f1 = f;
            f = f0 + f1;
        }
        return f;
    }
};

File: 1013-fibonacci-number/fibonacci-number.java
class Solution {
    public int fib(int n) {
        int[] dp = new int[n + 1];
        Arrays.fill(dp, -1);
        return fibo(dp, n);
    }

    public int fibo(int[] dp, int n){
        if (n == 0) return 0;
        if (n == 1) return 1;
        if (dp[n] != -1) return dp[n];
        dp[n] = fibo(dp, n - 1) + fibo(dp, n - 2);
        return dp[n];
    }
}